cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedSQLEngineCollection.java

```diff
@@ -51,35 +51,118 @@
 
   public WrappedSQLEngineCollection(SQLBackedCollection<T> wrapped,
                                     java.util.function.Function<SparkCollection<T>, SparkCollection<U>> mapper) {
-    this.wrapped = wrapped;
     this.mapper = mapper;
-  }
-
-  private SparkCollection<U> unwrap() {
-    if (unwrapped == null) {
-      unwrapped = mapper.apply(wrapped);
-    }
-
-    return unwrapped;
+    this.wrapped = wrapped;
   }
 
   /**
    * Executes an operation on the underlying collection and then wraps it in the same mapper for this collection.
    *
-   * This is useful when executing multiple operations in sequence where we need to delegate the operation to the
-   * underlying SQL engine and keep delaying the pull operation.
+   * This is useful when executing multiple operations in sequence where we need to delegate an operation to the
+   * underlying implementation
    *
    * By calling this over all wrapped collections, we will eventually reach an instance of a
    * {@link SQLEngineCollection} where the actual operation will take place.
    *
    * @param remapper function used to re-map the underlying collection.
    * @return SQL Backed collection after re-mapping the underlying colleciton and re-adding the mapper.
    */
-  private SparkCollection<U> rewrap(
+  protected SparkCollection<U> rewrap(
     java.util.function.Function<SparkCollection<T>, SparkCollection<T>> remapper) {
     return new WrappedSQLEngineCollection<>((SQLBackedCollection<T>) remapper.apply(wrapped), mapper);
   }
 
+  protected SparkCollection<U> unwrap() {
+    if (unwrapped == null) {
+      unwrapped = mapper.apply(wrapped);
+    }
+
+    return unwrapped;
+  }
+
+  @Override
+  public boolean tryStoreDirect(StageSpec stageSpec) {
```

Comment:
Please restore original method order. It would be much easier to see if anything changed.